### PR TITLE
Smart contract

### PR DESCRIPTION
**Pull request by twzjwang**

### CODE REVIEW

**Comment by jserv** [comment]
Can you append some materials for further work regarding blind signature?

File: paper.tex [hunk 1]
```diff
@@ -230,6 +231,26 @@ \subsubsection{Components}
...
+C = r^e m
+\end{equation}
+
+The broker signs the blinding message with his private key $d$.
+
+\begin{equation}
+S = C^d
+\end{equation}
+
+$S$ is the broker's signature of $C$. In RSA system, $ed$ is equal to 1. To remove the blind factor, the data provider computes the following calculation.
```
Comment (jkrvivian) [inline paper.tex]: blind factor => blinding factor

File: paper.tex [hunk 2]
```diff
@@ -369,6 +390,119 @@ \section{Results}
...
+			state == State.Launched,
+			"One key has been certified."
+		);
+		signedBlindedKey = _signedBlindedKey;
+		state = State.KeyCertified;
+		emit newSignedBlindedKey(signedBlindedKey);
+	}
+\end{lstlisting}
+
+Before a data provider asks a broker to certify a session key, he blinds the session key first. The blinded session key is published to the smart contract by calling $addBlindedKey()$. Also the broker should use $addSignedBlindedKey()$ to publish the signature. As a result, the key certification process is recorded on the smart contract which is traceable afterward. By the way, it is hard for an attacker to compute the session key by blinding messages and signatures on smart contracts.
```
Comment (jkrvivian) [inline paper.tex]: hard for an attacker to compute the session key `by` => `with`

File: paper.tex [hunk 3]
```diff
@@ -369,6 +390,119 @@ \section{Results}
...
+		update encrypt key & Provider & 201624 & 166784 \\ \hline
+		update number & Broker & 54289 & 32825 \\ \hline
+		ask refund & Consumer & 43817 & 22545 \\ \hline
+		withdraw & Broker/Provider/Consumer & 35991 & 14719 \\ \hline
+	\end{tabular}
+\end{table}
+
+Table \ref{tab:gas} shows the gas usage for each function. Creating a smart contract (Listing \ref{lst:constructor} ) costs the most gas, and functions (Listing \ref{lst:key_certification} and Listing \ref{lst:key_exchange}) that are related to key certification and key exchange need a lot of gas as well. Storage on blockchain costs a lot. It is expensive to use memory in Solidity. For instance, an SSTORE operation may cost 20000 gas when the value is set to non-zero from zero\cite{Ethereum} according to Ethereum Yellow Paper. Smart contract initialization, storing ciphertexts and signatures during key certification and key exchange needs a lot of operations and storages which causes high gas consumption. On the other hand, we can observe that the broker is responsible for the most part of transaction fees while consumers only need to send one transaction to subscribe the product if they don't have to refund.
```
Comment (jkrvivian) [inline paper.tex]: \cite{Ethereum} should be used in `Ethereum Yellow Paper`

**Comment by jserv** [comment]
@twzjwang, any further update upon requested?

**Pull request merged**